Use step latents for reference style

jasonphillips opened the pull request:
This is an attempt to solve the problem I identified in:  https://github.com/brianfitzgerald/style_aligned_comfy/issues/5#issuecomment-1875971490

I believe the primary issue with the current implementation was that it doesn't send reference latents for every step. I'm getting much better results with this setup, even when the prompt doesn't capture the style's specifics well (I think the demos in this repo use prompts so aligned with the source style that they give a slightly too optimistic impression of how well the style referencing is actually working).

I'll attach some example workflows in the comments.

Changes
 - added new node that creates the noised latents for the reference image and passes all of the step-aligned latents forward; this replaces having to use a sampler manually with flipped sigmas
 - use this output to replace the reference latent with the correct one after each step, to keep the ref latent from denoising away from its target
 - also adjusted the precision of the `scale` parameter, because I've found that small changes at the `.01` level make all the difference. 
 
 [This is where](https://github.com/google/style-aligned/blob/2d4283a42b656ef0827fae5d5b616d51ba3704f4/inversion.py#L111-L117) they follow the above strategy in the paper's official code, replacing the reference latent at every step during inference rather than relying only on the final noised DDIM inversion for the trajectory. 
 
 I think there are still some challenges in that the noise probably isn't being generated as well with this approach as it is in the paper's implementation, ~plus ComfyUI has the annoying limitation currently that we can't pass a different prompt for different samples in the same pass during sampling~ 
 
 Edit -- I've added a commit that separates the conditioning, which significantly improves the prompt alignment.

jasonphillips commented:
Here's one of my results for the painting with the robot, following the example in the repo; I think it comes out much better.
![robot5](https://github.com/brianfitzgerald/style_aligned_comfy/assets/5672408/012249ee-9da5-4818-89b2-542f93bbe2e6)




And attached below are workflows for rendering a robot using this reference image:
![papercraft](https://github.com/brianfitzgerald/style_aligned_comfy/assets/5672408/bcefb033-5230-4a61-a13d-d6e2e58fd0c1)
  
  

**First, with only queries and keys shared:**
workflow: [robot_paper1.json](https://github.com/brianfitzgerald/style_aligned_comfy/files/13836387/robot_paper1.json)
![robot_paper1](https://github.com/brianfitzgerald/style_aligned_comfy/assets/5672408/9a0c2542-c47b-4bdf-a324-ee67cf88854a)

**Second, with queries+keys+values shared:**
workflow: [workflow_robotpaperqkv.json](https://github.com/brianfitzgerald/style_aligned_comfy/files/13836388/workflow_robotpaperqkv.json)
![robot_qkv](https://github.com/brianfitzgerald/style_aligned_comfy/assets/5672408/4a9a5f34-cb16-4c17-891d-341e0a6abfb9)

GXcells commented:
I confirm that it works pretty well.
I think the workflow could be cleaned a bit if merging everything in only 1 node.

Seems to work much better with SDXL compared to SD1.5

jasonphillips commented:
@GXcells Thanks for testing it out, I'm glad to hear from others. 

I've similarly noticed that the CFG of the reference encoding and of the main sampling makes a huge difference -- in fact I can make the biggest impact on quality of the output by playing with these 2 numbers, often even more than using the "scale" parameter itself. Finding the sweet spot for each image task is the challenge in each case, but with enough time I usually find some values that work. 

I think there are 2 limitations involved here:
  1. the encoding part is basically a "DDIM inversion" task, and making a stable inversion trajectory is still an area of active research, as the paper authors note in their appendix. They take one simple approach in their official code, and we could try to adapt that directly here, but for the moment we're relying on just a normal backwards loop through Comfy's sampling and I'm not fully knowledgeable yet about how well that handles it since some parts of Comfy's stack aren't 100% clear to me. For context, this is essentially the process we're looking at when we attempt a CFG-driven inversion: https://arxiv.org/pdf/2211.09794.pdf
  2. Comfy doesn't let us use a different prompt for different parts of the same batch -- so during the main sampling loop, the reference latents are being attended to by the context/prompt of the _target_ image only, and that seems suboptimal.

jasonphillips commented:
(small additional note -- you may want to try setting the "scale" to _below_ 1 (eg. 0.85), depending on the task. That's generally how the paper authors used it, but in my example workflows above I was setting it higher to try and emphasize the style impact.

jasonphillips commented on __init__.py:
```diff
@@ -222,7 +271,13 @@ def patch(
+        def callback(step: int, x0: T, x: T, steps: int):
+            preview_callback(step, x0, x, steps)
+            if (step < steps):
+                x[0] = ref_latents[step]
+                x0[0] = ref_latents[step]
```
The reason I overwrite both the output and the reference sample is that this callback is apparently invoked by Comfy _before_ the sampling step subtracts the predicted output from the sample as the predicted noise, and uses that to make its next adjustment based on the particular sampler algorithm chosen; so we want the predicted noise to be 0, and avoid any _additional_ adjustment after we manually set the desired step output.

jasonphillips commented:
I've now also adjusted it to separate the conditioning for the reference image from the conditioning for the rest of the batch; this gives more stable results and follows more closely the official paper code. It does add a new required input to the sampler, which is the positive conditioning/prompt for the reference image.

Updated workflow examples:
![workflow_truck2](https://github.com/brianfitzgerald/style_aligned_comfy/assets/5672408/58da49ad-db95-4eb4-aec6-85a01d9e7a41)

Ref image used: 
<img src="https://github.com/brianfitzgerald/style_aligned_comfy/assets/5672408/6903813e-cba3-4c06-89bc-3e1bdc23dbae" width="300"/>



![workflow_plane](https://github.com/brianfitzgerald/style_aligned_comfy/assets/5672408/4b53d0fe-52a7-4a86-854c-63d6b8b5b39f)
ref image used: 
<img src="https://github.com/brianfitzgerald/style_aligned_comfy/assets/5672408/411edd8e-55b6-4147-a494-00ce0d9a60d5" width="300"/>

jasonphillips commented:
Here's a workflow using Controlnet, and SD 1.5 instead of XL.

A few things I've noticed:
 - ancestral samplers like `euler a` can work extremely well; I suspect the added noise helps correct for any weirdness in the inverse DDIM trajectory
 - playing with the CFG matters
 - scale can be useful if you make small adjustments

![workflow_barn](https://github.com/brianfitzgerald/style_aligned_comfy/assets/5672408/11c68e93-33e8-43b8-b985-edf33cfc07a9)

GXcells commented:
> Here's a workflow using Controlnet, and SD 1.5 instead of XL.
> 
> A few things I've noticed:
> 
> * ancestral samplers like `euler a` can work extremely well; I suspect the added noise helps correct for any weirdness in the inverse DDIM trajectory
> * playing with the CFG matters
> * scale can be useful if you make small adjustments
> 
> ![workflow_barn](https://private-user-images.githubusercontent.com/5672408/295041784-11c68e93-33e8-43b8-b985-edf33cfc07a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ3NTA0NDUsIm5iZiI6MTcwNDc1MDE0NSwicGF0aCI6Ii81NjcyNDA4LzI5NTA0MTc4NC0xMWM2OGU5My0zM2U4LTQzYjgtYjk4NS1lZGYzM2NmYzA3YTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDhUMjE0MjI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTBlYTk5YmIyZTk3YzY1ODczZmUzNGFlNzlkYTcyNDA3Y2Y1MGE0ZDc1NmI1MzQ4MzQxNWZmODFkNjdkZjRhZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Sva8mCi806k1bjT3vg4LzDRW1_qDTu8Vs5JJwUfbqY0)

Wow pretty impressive! I love that you used control net to change the style. This is a pretty damn good style transfer workflow for img2img!!!

jasonphillips commented:
I pushed one more commit, which gives much greater stability with controlnet. The change is to allow you to send a corresponding controlnet condition through the reference positive condition, which will then be batched along with the main condition correctly so that the reference image uses its controlnet context and the generated images use theirs. 

Without this change, controlnet was applying its context to both the reference and the new images during generation, and this caused some of the instability. 

I've updated my barn example which has _much_ easier and better coherence this way (applying controlnet to both reference condition and generated condition) and is no longer as volatile with CFG etc as I was experiencing before. 

![workflow_barn_new](https://github.com/brianfitzgerald/style_aligned_comfy/assets/5672408/8d48fabe-0acc-4298-acf3-c7f9f4505754)

GXcells commented:
It seems that the style aligned effect completely breaks (is completely lost) if I use a dimension (height or width) higher than 768 pixels.

Any hints?

jasonphillips commented:
> It seems that the style aligned effect completely breaks (is completely lost) if I use a dimension (height or width) higher than 768 pixels.
> 
> Any hints?

I haven't seen that problem -- I'm alternating 768 and 1024 for different tasks and so far both have worked well. Try to pull my latest code here just in case, since it has some minor tweaks. 

If you post a workflow I could try it myself and see.

GXcells commented:
> If you post a workflow I could try it myself and see

Here is one workflow (using sdxl turbo but I get same result with sdxl). Also I use --lowVRAM argument as I only have 4GB VRAM

All good if input is 768*768 but if I increase size it breaks
![ComfyUI_temp_jmylp_00001_](https://github.com/brianfitzgerald/style_aligned_comfy/assets/152809571/4d75eb68-761c-4287-829d-374cbb7dd6e1)

jasonphillips commented:
> 
> Here is one workflow (using sdxl turbo but I get same result with sdxl). Also I use --lowVRAM argument as I only have 4GB VRAM
> 
> All good if input is 768*768 but if I increase size it breaks ![ComfyUI_temp_jmylp_00001_](https://private-user-images.githubusercontent.com/152809571/296055736-4d75eb68-761c-4287-829d-374cbb7dd6e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUwMjU2MjMsIm5iZiI6MTcwNTAyNTMyMywicGF0aCI6Ii8xNTI4MDk1NzEvMjk2MDU1NzM2LTRkNzVlYjY4LTc2MWMtNDI4Ny04MjlkLTM3NGNiYjdkZDZlMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDExMlQwMjA4NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lYjIxYWM0N2JhM2YyMzYxNTkzMjhhNGZjZGM2MWEyNDQ1MmQ5MzkyMzI3MzgzYWEyNWJhMjJlN2IwY2QzZGVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.uZN803naa_1acoskm8rR1s5EGe9SJMykJy5wxg57lXg)

Oh you're using SDXL Turbo -- I haven't tried that for this style aligned technique, but I wonder if it will work very well since it's tuned for using very few steps, and the style alignment seems to work better with many steps (as I recall, all the pretty examples from the paper authors were generated with _50_ steps). 

I get very nice outputs for your workflow using regular SDXL, 30 steps, higher cfg (8.0), and switching to euler ancestral sampler. But I haven't had a chance to try out Turbo with it.

GXcells commented:
> Oh you're using SDXL Turbo

Yes but it is same if I use SDXL normal with higher steps and CFG.

I think there is something wrong with the --low VRAM argument that I use to run comfyUI which affects something under the hood for images that have more than 768 pixels in one dimension.

bmaltais commented:
This should be merged to the main branch... not sure why it has not been merged yet...

bmaltais commented:
@jasonphillips I have "improved"... maybe your workflow by adding a bit of IPAdapter to it... I wish I could share the workflow screen captures like you do but I don't know how... so here is the closest I know how to do... I will share the output image created by the workflow instead:

![image](https://github.com/brianfitzgerald/style_aligned_comfy/assets/7474674/6f1bf32e-ef90-4870-83d7-ee7c092cf3cb)

And here is the reference image I used for it:

![image](https://github.com/brianfitzgerald/style_aligned_comfy/assets/7474674/3cb24cbc-9d71-442f-a999-9be3d311250b)

Somehow this only work with the base SDXL 1.0 model... not with SD1.5... not sure why...

bmaltais commented:
Here is another example of using the same workflow... but with a Monet painting instead as a reference:

Workflow:

[workflow monet.json](https://github.com/brianfitzgerald/style_aligned_comfy/files/13944856/workflow.monet.json)

Output:

![image](https://github.com/brianfitzgerald/style_aligned_comfy/assets/7474674/a589c41e-8c0b-443f-94ae-5c6875b397e8)

Reference:

![image](https://github.com/brianfitzgerald/style_aligned_comfy/assets/7474674/00432e9c-e5f6-4fb3-92d5-c364b047b5c9)

bmaltais commented:
I also discovered that using a denoise of 0.5 (instead of 1.0) for StyleAligned Reference with a CFG of 7 (instead of 4) produce results closer to the reference.

dfl commented:
> I also collapsed a lot of the nodes into "workflow" nodes.
@bmaltais how I can I run this workflow? Not sure how to access "workflow/Style Aligned Reference"

bmaltais commented:
> > I also collapsed a lot of the nodes into "workflow" nodes.
> > @bmaltais how I can I run this workflow? Not sure how to access "workflow/Style Aligned Reference"

It should work out of the box. Make sure to update comfy to the latest release. Also, you need to manually install the version with the PR otherwise you will not have the right nodes.

jasonphillips commented:
@bmaltais 

Nice, good to see other results -- I haven't tried it with the ipadapter but I still get good results on your workflow even just removing that piece and using only the style node. 

As for SD 1.5, I've found that a few things can make a big difference:
 - simpler schedule (DDIM uniform etc) and simpler sampler (euler ancestral is pretty good)
 - low CFG as you said, but not always (sometimes I try higher and that suddenly works better... effect seems more unstable on 1.5 in general)

>  I wish I could share the workflow screen captures like you do but I don't know how... so here is the closest I know how to do... 

I use `ComfyUI-Custom-Scripts` which has a feature for exporting your entire workflow as an image:  https://github.com/pythongosssss/ComfyUI-Custom-Scripts?tab=readme-ov-file#workflow-images

jesenzhang commented:
@jasonphillips  
Got size error, please help！
![image](https://github.com/brianfitzgerald/style_aligned_comfy/assets/7984556/7c061a43-ab8c-4d92-a70f-91b6d6bd5a1a)

jasonphillips commented:
> @jasonphillips Got size error, please help！ ![image](https://private-user-images.githubusercontent.com/7984556/297650635-7c061a43-ab8c-4d92-a70f-91b6d6bd5a1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU1OTE2ODksIm5iZiI6MTcwNTU5MTM4OSwicGF0aCI6Ii83OTg0NTU2LzI5NzY1MDYzNS03YzA2MWE0My1hYjhjLTRkOTItYTcwZi05MWI2ZDZiZDVhMWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMThUMTUyMzA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2VhMmE1ZWEwMjFlNTE0NmJhY2Q2YzUzYWM5ZmFmYmMzNDFjNzhjNWU0OWVkMTIzZjAzMzgwYjRmOWM1NDgwYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.pyzkUmw8AzuWpGkDcF64LDfwzGflsspa6Wdj7_ImuUA)

Could you post your workflow? That error indicates that the conditioning (positive input) is 2x (usually it's 77 in SD models, but yours is double) and is therefore larger than the reference conditioning input. I could try to just zero-pad in that case but I'm curious why you'd have a different dimension here, possibly it's something else in your pipeline or using a tool to enable very long prompts.

jesenzhang commented:
> > @jasonphillips Got size error, please help！ ![image](https://private-user-images.githubusercontent.com/7984556/297650635-7c061a43-ab8c-4d92-a70f-91b6d6bd5a1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU1OTE2ODksIm5iZiI6MTcwNTU5MTM4OSwicGF0aCI6Ii83OTg0NTU2LzI5NzY1MDYzNS03YzA2MWE0My1hYjhjLTRkOTItYTcwZi05MWI2ZDZiZDVhMWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMThUMTUyMzA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2VhMmE1ZWEwMjFlNTE0NmJhY2Q2YzUzYWM5ZmFmYmMzNDFjNzhjNWU0OWVkMTIzZjAzMzgwYjRmOWM1NDgwYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.pyzkUmw8AzuWpGkDcF64LDfwzGflsspa6Wdj7_ImuUA)
> 
> Could you post your workflow? That error indicates that the conditioning (positive input) is 2x (usually it's 77 in SD models, but yours is double) and is therefore larger than the reference conditioning input. I could try to just zero-pad in that case but I'm curious why you'd have a different dimension here, possibly it's something else in your pipeline or using a tool to enable very long prompts.

This is the workflow,it is ok when using the same positive and negative conditions  for the stylealigen group.
![workflow_bug](https://github.com/brianfitzgerald/style_aligned_comfy/assets/7984556/e96419d1-a15b-479c-a17a-b15daee8f9ab)

jesenzhang commented:
> > @jasonphillips Got size error, please help！ ![image](https://private-user-images.githubusercontent.com/7984556/297650635-7c061a43-ab8c-4d92-a70f-91b6d6bd5a1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU1OTE2ODksIm5iZiI6MTcwNTU5MTM4OSwicGF0aCI6Ii83OTg0NTU2LzI5NzY1MDYzNS03YzA2MWE0My1hYjhjLTRkOTItYTcwZi05MWI2ZDZiZDVhMWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMThUMTUyMzA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2VhMmE1ZWEwMjFlNTE0NmJhY2Q2YzUzYWM5ZmFmYmMzNDFjNzhjNWU0OWVkMTIzZjAzMzgwYjRmOWM1NDgwYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.pyzkUmw8AzuWpGkDcF64LDfwzGflsspa6Wdj7_ImuUA)
> 
> Could you post your workflow? That error indicates that the conditioning (positive input) is 2x (usually it's 77 in SD models, but yours is double) and is therefore larger than the reference conditioning input. I could try to just zero-pad in that case but I'm curious why you'd have a different dimension here, possibly it's something else in your pipeline or using a tool to enable very long prompts.

This is the workflow,it is ok when using the same positive and negative conditions  for the stylealigned group.
![workflow_bug](https://github.com/brianfitzgerald/style_aligned_comfy/assets/7984556/6a6fd87e-9c3b-43e4-95b3-ebb02dee806e)

jasonphillips commented:
> > > @jasonphillips Got size error, please help！ ![image](https://private-user-images.githubusercontent.com/7984556/297650635-7c061a43-ab8c-4d92-a70f-91b6d6bd5a1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU1OTE2ODksIm5iZiI6MTcwNTU5MTM4OSwicGF0aCI6Ii83OTg0NTU2LzI5NzY1MDYzNS03YzA2MWE0My1hYjhjLTRkOTItYTcwZi05MWI2ZDZiZDVhMWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMThUMTUyMzA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2VhMmE1ZWEwMjFlNTE0NmJhY2Q2YzUzYWM5ZmFmYmMzNDFjNzhjNWU0OWVkMTIzZjAzMzgwYjRmOWM1NDgwYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.pyzkUmw8AzuWpGkDcF64LDfwzGflsspa6Wdj7_ImuUA)
> > 
> > 
> > Could you post your workflow? That error indicates that the conditioning (positive input) is 2x (usually it's 77 in SD models, but yours is double) and is therefore larger than the reference conditioning input. I could try to just zero-pad in that case but I'm curious why you'd have a different dimension here, possibly it's something else in your pipeline or using a tool to enable very long prompts.
> 
> This is the workflow,it is ok when using the same positive and negative conditions for the stylealigned group. ![workflow_bug](https://private-user-images.githubusercontent.com/7984556/297964096-6a6fd87e-9c3b-43e4-95b3-ebb02dee806e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU3MDQyMDUsIm5iZiI6MTcwNTcwMzkwNSwicGF0aCI6Ii83OTg0NTU2LzI5Nzk2NDA5Ni02YTZmZDg3ZS05YzNiLTQzZTQtOTViMy1lYmIwMmRlZTgwNmUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMTlUMjIzODI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTE5Yjc4YmY5YjA2ODFkNmY1ZDQ0ODBmNmQxOWU0N2RhYWUyNWIzNWE4M2Y2YmE4ZWRjMjEyMTkwN2MyOWZiOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.dvyjx9NAHl3msGIuFfCW6QW9zeeoOMyb3SoJhKz6xP4)

That's a lot of custom nodes I don't have! 

At a glance, it's definitely caused by your condition for the positive having a different shape from the ref_positive (something in your complex prompting perhaps is making it use a 2x prompt input size), and this isn't fully handled by the code. I guess I should pad with zeros when creating the batch conditions in order to solve this.

brianfitzgerald commented:
LGTM - I'll deprecate the old reference sampler and replace it with this node, but I see no harm in merging now as it looks like a clear upgrade over the older reference sampler. Docs and example workflows will need to be updated - feel free to submit a PR to do so, but if you don't I'll get to it in a few days. Thanks for submitting!
